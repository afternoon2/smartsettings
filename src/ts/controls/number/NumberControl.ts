import { Control } from '../control/Control';
import {
  NumberControlProps, ControlListener, ControlListenerUpdate,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/number.sass';

export class NumberControl extends Control {
  public controlElement: HTMLInputElement;

  protected static template = (state: InternalState): string => `<input
      type="number"
      class="${Styles.number}"
      ${state.value ? `value="${state.value}"` : ''}
      ${typeof state.min === 'number' ? `min="${state.min}"` : ''}
      ${typeof state.max === 'number' ? `max="${state.max}"` : ''}
      ${typeof state.step === 'number' ? `step="${state.step}"` : ''}
      ${state.readonly ? 'readonly="true"' : ''}
      ${state.disabled ? 'disabled' : ''}
      ${state.placeholder ? `placeholder="${state.placeholder}"` : ''}
    />
  `;

  constructor(props: NumberControlProps) {
    super(props, NumberControl.template);
    this.controlElement = this.element.querySelector('[type="number"]') as HTMLInputElement;
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: number) {
    this.state.value = value;
  }

  get value(): number {
    return this.state.value as number;
  }

  private onValue: ControlListener = (update: ControlListenerUpdate) => {
    this.controlElement.value = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('change', (e: Event) => {
      this.value = parseFloat((e.target as HTMLInputElement).value);
    });
  }
}