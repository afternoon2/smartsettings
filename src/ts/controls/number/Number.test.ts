import { NumberControl } from './NumberControl';

let control: NumberControl;

const userListener = jest.fn();

describe('Number control', () => {
  beforeEach(() => {
    control = new NumberControl({
      id: 'id',
      name: 'name',
      options: {
        value: 0,
        min: 0,
        max: 10,
        step: 0.1,
        disabled: false,
      },
      parentElement: document.body,
      userListener,
    });
  });

  test('If it matches snapshot after initial load', () => {
    expect(control.element).toMatchSnapshot();
  });

  test('Disable', () => {
    control.disable();
    expect(control.element).toMatchSnapshot();
  });

  test('Enable', () => {
    control.disable();
    control.enable();
    expect(control.controlElement.hasAttribute('disabled')).toBe(false);
    expect(control.element).toMatchSnapshot();
  });

  test('Get and set value', () => {
    control.value = 4;
    expect(control.value).toBe(4);
  });

  test('If it assings user listener correctly', () => {
    expect(userListener).toHaveBeenCalledTimes(4);
  });
});