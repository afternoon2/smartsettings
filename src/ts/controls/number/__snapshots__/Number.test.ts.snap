// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number control Disable 1`] = `
<div
  class="control disabled"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="name"
    >
      name
    </p>
    <div
      class="control__content"
    >
      <input
        class="number"
        disabled="true"
        max="10"
        step="0.1"
        type="number"
      />
      
  
    </div>
  </main>
</div>
`;

exports[`Number control Enable 1`] = `
<div
  class="control"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="name"
    >
      name
    </p>
    <div
      class="control__content"
    >
      <input
        class="number"
        max="10"
        step="0.1"
        type="number"
      />
      
  
    </div>
  </main>
</div>
`;

exports[`Number control If it matches snapshot after initial load 1`] = `
<div
  class="control"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="name"
    >
      name
    </p>
    <div
      class="control__content"
    >
      <input
        class="number"
        max="10"
        step="0.1"
        type="number"
      />
      
  
    </div>
  </main>
</div>
`;
